File: python/z3t_ai_agent/llm.py
from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, Any

from .types import ResolvedConfig

if TYPE_CHECKING:
    from anthropic import Anthropic
    from google.genai import Client as GoogleClient
    from openai import OpenAI


@dataclass
class LlmClients:
    """Pre-configured LLM provider clients pointing at the z3t LLM proxy.

    Each is `None` if the corresponding optional dependency isn't installed —
    developers who don't call a given `ctx.llm.*` client don't need that
    provider's package installed at all.
    """

    openai: "OpenAI | None"
    anthropic: "Anthropic | None"
    google: "GoogleClient | None"


def create_llm_clients(config: ResolvedConfig, call_id: str | None = None) -> LlmClients:
    """Build a fresh set of LLM clients for a single call, so `x-agent-call-id` can be
    attributed correctly. Cheap — this only wraps configuration, it doesn't open
    a connection."""
    headers = {"x-agent-call-id": call_id} if call_id else {}

    openai_client: Any = None
    try:
        from openai import OpenAI

        openai_client = OpenAI(
            base_url=f"{config.base_url}/llm/openai/v1",
            api_key=config.api_key,
            default_headers=headers,
        )
    except ImportError:
        pass

    anthropic_client: Any = None
    try:
        from anthropic import Anthropic

        anthropic_client = Anthropic(
            base_url=f"{config.base_url}/llm/anthropic",
            api_key=config.api_key,
            default_headers=headers,
        )
    except ImportError:
        pass

    google_client: Any = None
    try:
        from google import genai
        from google.genai import types as genai_types

        # Unlike the TypeScript SDK's Z3tGoogleGenerativeAI wrapper (which only injects
        # baseUrl and can't carry the call-id header — see CLAUDE.md), google-genai's
        # Client takes both base_url and headers directly via HttpOptions, so Google
        # calls here are attributable to a specific call same as OpenAI/Anthropic.
        google_client = genai.Client(
            api_key=config.api_key,
            http_options=genai_types.HttpOptions(
                base_url=f"{config.base_url}/llm/google",
                headers=headers,
            ),
        )
    except ImportError:
        pass

    return LlmClients(openai=openai_client, anthropic=anthropic_client, google=google_client)

File: typescript/src/llm.ts
import type OpenAI from 'openai'
import type Anthropic from '@anthropic-ai/sdk'
import type { GoogleGenAI } from '@google/genai'
import type { ResolvedConfig } from './types'

export type LlmClients = {
  openai: OpenAI
  anthropic: Anthropic
  google: GoogleGenAI
}

// Lazy getters — each LLM package is only require()'d when the handler actually
// accesses ctx.llm.openai / .anthropic / .google, so missing optional peer deps
// don't crash agents that don't use that provider.
export function createLlmClients(config: ResolvedConfig, callId?: string): LlmClients {
  const callHeaders = callId ? { 'x-agent-call-id': callId } : {}
  let _openai: OpenAI | undefined
  let _anthropic: Anthropic | undefined
  let _google: GoogleGenAI | undefined

  return {
    get openai() {
      if (!_openai) {
        try {
          // eslint-disable-next-line @typescript-eslint/no-require-imports
          const { default: Ctor } = require('openai') as { default: new (opts: object) => OpenAI }
          _openai = new Ctor({ baseURL: `${config.baseUrl}/llm/openai/v1`, apiKey: config.apiKey, defaultHeaders: callHeaders })
        } catch {
          throw new Error("ctx.llm.openai requires the 'openai' package. Run: npm install openai")
        }
      }
      return _openai
    },
    get anthropic() {
      if (!_anthropic) {
        try {
          // eslint-disable-next-line @typescript-eslint/no-require-imports
          const { default: Ctor } = require('@anthropic-ai/sdk') as { default: new (opts: object) => Anthropic }
          _anthropic = new Ctor({ baseURL: `${config.baseUrl}/llm/anthropic`, apiKey: config.apiKey, defaultHeaders: callHeaders })
        } catch {
          throw new Error("ctx.llm.anthropic requires the '@anthropic-ai/sdk' package. Run: npm install @anthropic-ai/sdk")
        }
      }
      return _anthropic
    },
    get google() {
      if (!_google) {
        try {
          // eslint-disable-next-line @typescript-eslint/no-require-imports
          const { GoogleGenAI: Ctor } = require('@google/genai') as { GoogleGenAI: new (opts: object) => GoogleGenAI }
          _google = new Ctor({
            apiKey: config.apiKey,
            httpOptions: { baseUrl: `${config.baseUrl}/llm/google`, headers: callHeaders },
          })
        } catch {
          throw new Error("ctx.llm.google requires the '@google/genai' package. Run: npm install @google/genai")
        }
      }
      return _google
    },
  }
}
